service/history/queueTaskProcessor.go

```diff
@@ -0,0 +1,201 @@
+// THE SOFTWARE.
+
+package history
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/task"
+)
+
+type (
+	queueTaskProcessorOptions struct {
+		fifoSchedulerOptions *task.FIFOTaskSchedulerOptions
+		wRRSchedulerOptions  *task.WeightedRoundRobinTaskSchedulerOptions
+	}
+
+	queueTaskProcessorImpl struct {
+		sync.RWMutex
+
+		priorityAssigner taskPriorityAssigner
+		schedulers       map[int]task.Scheduler
+
+		status        int32
+		options       *queueTaskProcessorOptions
+		logger        log.Logger
+		metricsClient metrics.Client
+
+		// TODO: add a host level task scheduler
+	}
+)
+
+var (
+	errTaskProcessorNotRunning = errors.New("queue task processor is not running")
+)
+
+func newQueueTaskProcessor(
+	priorityAssigner taskPriorityAssigner,
+	options *queueTaskProcessorOptions,
+	logger log.Logger,
+	metricsClient metrics.Client,
+) (queueTaskProcessor, error) {
+	if (options.fifoSchedulerOptions != nil && options.wRRSchedulerOptions != nil) ||
+		(options.fifoSchedulerOptions == nil && options.wRRSchedulerOptions == nil) {
+		return nil, errors.New("please specify options for exactly one type of task scheduler")
+	}
+	return &queueTaskProcessorImpl{
+		priorityAssigner: priorityAssigner,
+		schedulers:       make(map[int]task.Scheduler),
+		status:           common.DaemonStatusInitialized,
+		options:          options,
+		logger:           logger,
+		metricsClient:    metricsClient,
+	}, nil
+}
+
+func (p *queueTaskProcessorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.logger.Info("Queue task processor started.")
+}
+
+func (p *queueTaskProcessorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	p.Lock()
+	defer p.Unlock()
+
+	for shardID, scheduler := range p.schedulers {
+		delete(p.schedulers, shardID)
+		scheduler.Stop()
+	}
+
+	p.logger.Info("Queue task processor stopped.")
+}
+
+func (p *queueTaskProcessorImpl) StopShardProcessor(
+	shardID int,
+) {
+	p.Lock()
+	scheduler, ok := p.schedulers[shardID]
+	if !ok {
+		p.Unlock()
+		return
+	}
+
+	delete(p.schedulers, shardID)
+	p.Unlock()
+
+	// don't hold the lock while stopping the scheduler
+	scheduler.Stop()
+}
+
+func (p *queueTaskProcessorImpl) Submit(
+	task queueTask,
+) error {
+	scheduler, err := p.prepareSubmit(task)
+	if err != nil {
+		return err
+	}
+	return scheduler.Submit(task)
+}
+
+func (p *queueTaskProcessorImpl) TrySubmit(
+	task queueTask,
+) (bool, error) {
+	scheduler, err := p.prepareSubmit(task)
+	if err != nil {
+		return false, err
+	}
+	return scheduler.TrySubmit(task)
+}
+
+func (p *queueTaskProcessorImpl) prepareSubmit(
+	task queueTask,
+) (task.Scheduler, error) {
+	if !p.isRunning() {
```

Comment:
As discussed, this check could move to getOrCreateTaskScheduler

Reply:
Done.